companion/src/simulation/telemetryprovidercrossfire.cpp

```diff
@@ -0,0 +1,530 @@
+/*
+ * Copyright (C) EdgeTX
+ *
+ * Based on code named
+ *   opentx - https://github.com/opentx/opentx
+ *   th9x - http://code.google.com/p/th9x
+ *   er9x - http://code.google.com/p/er9x
+ *   gruvin9x - http://code.google.com/p/gruvin9x
+ *
+ * License GPLv2: http://www.gnu.org/licenses/gpl-2.0.html
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <QMessageBox>
+#include <QFileDialog>
+
+#include "appdata.h"
+#include "telemetryprovidercrossfire.h"
+#include "ui_telemetryprovidercrossfire.h"
+#include "telem_data.h"
+
+template<class t> t LIMIT(t mi, t x, t ma) { return std::min(std::max(mi, x), ma); }
+
+TelemetryProviderCrossfire::TelemetryProviderCrossfire(QWidget * parent):
+  QWidget(parent),
+  ui(new Ui::TelemetryProviderCrossfire)
+{
+  ui->setupUi(this);
+
+  // Set default values from UI definition into GPS
+  gps.setLatLon(ui->input_gps->text());
+  gps.setCourseDegrees(ui->input_hdg->value());
+  gps.setSpeedKMH(ui->input_gspd->value());
+  gps.setSatelliteCount(ui->input_sats->value());
+  gps.setAltitude(ui->input_alt->value());
+
+  connect(ui->input_hdg,  SIGNAL(valueChanged(double)),
+          &gps,           SLOT(setCourseDegrees(double)));
+  connect(ui->input_gps,  &QLineEdit::textChanged,
+          &gps,           &SimulatedGPS::setLatLon);
+  connect(ui->input_gspd, SIGNAL(valueChanged(double)),
+          &gps,           SLOT(setSpeedKMH(double)));
+  connect(ui->input_alt,  SIGNAL(valueChanged(double)),
+          &gps,           SLOT(setAltitude(double)));
+  connect(ui->input_sats, SIGNAL(valueChanged(int)),
+          &gps,           SLOT(setSatelliteCount(int)));
+
+  connect(&gps,           &SimulatedGPS::positionChanged,
+          ui->input_gps,  &QLineEdit::setText);
+  connect(&gps,           &SimulatedGPS::courseDegreesChanged,
+          [this](double v) { ui->input_hdg->setValue(v); });
+
```

Comment:
For code consistency with existing code please keep Qt::connect functions to a single line. We usually only split across lines when lambda functions used for the receiver.
Please replace the use of SIGNAL and SLOT macros with function pointers and lambda functions. The macros do not provide compile time checking of the sender and receiver events.